langkit/dsl.py

```diff
@@ -656,186 +795,49 @@ def has_abstract_list(cls):
-
-        # Make Alternative instances available as EnumNode class attributes for
-        # a convenient way to create parsers for them.
-        for alt in alts:
-            attr_name = (names.Name('alt') + alt.name).lower
-            dct[attr_name] = alt
-
-        cls = type.__new__(mcs, name, bases, dct)
-
-        mcs.enum_types.append(cls)
-        for alt in alts:
-            alt._enum_node_cls = cls
-
-        return cls
-
-
-class EnumNode(BaseStruct):
-    """
-    Using this base class, users can create a hierarchy of nodes deriving from
-    the root node that are similar to an enum type. By declaring an EnumNode
-    derived type in the following way::
-
-        class Foo(T.EnumNode):
-            alternatives = ['bar', 'baz']
-
-    The user will get:
-
-    * An abstract node type the base type, deriving from T.root_node, denoted
-      by Foo.
-    * A concrete but empty node type for every alternative of the enum type,
-      that can be denoted by Foo.alt_{alt_name}.
-
-    Instead of providing explicit alternatives, the user can just define the
-    EnumNode as a qualifier node::
-
-        class Foo(T.EnumNode):
-            qualifier = True
-
-    In which case, alternatives will automatically be "present" and "absent",
-    and an as_bool method will be automatically generated.
-    """
-
-    __metaclass__ = _EnumNodeMetaclass
-
-    _alternatives = None
-    """
-    List of alternatives: one for each enum value.
-    :type: list[EnumNode.Alternative]
-    """
-
-    _qualifier = None
-    """
-    If provided and set to True, the alternatives used are "present" and
-    "absent".
-
-    :type: bool
-    """
-
-    _alt_present = None
-    _alt_absent = None
-    """
-    If ``_qualifier`` is true, direct references to both alternatives.
-
-    :type: EnumNode.Alternative
-    """
-
-    class Alternative(object):
-        def __init__(self, name):
-            """
-            :param names.Name name: Alternative name.
-            """
-            self.name = name
-
-            self._enum_node_cls = None
-            self._type = None
-
-        @property
-        def enum_node_cls(self):
-            """
-            EnumNode subclass that this alternative belongs to.
-            """
-            assert self._enum_node_cls
-            return self._enum_node_cls
-
-        @property
-        def type(self):
-            """
-            ASTNodeType subclass corresponding to this alternative.
-            """
-            assert self._type
-            return self._type
-
-        @property
-        def type_ref(self):
-            """
-            TypeRepo.Defer instance for the ASTNodeType subclass corresponding
-            to this alternative.
-            """
-            def get():
-                assert self._type
-                return self._type
-            return T.Defer(get, 'Alternative({})'.format(self.name.camel))
+class EnumNodeAlternative(object):
```

Comment:
This class is not supposed to be used outside of the Langkit engine (i.e. it’s not part of the DSL), so let’s rename it appropriately (e.g. _EnumNodeAlternative)